docs/SIL.rst

```diff
@@ -2608,6 +2660,76 @@ undefined behavior if the global variable has already been initialized.
 
 The type operand must be a lowered object type.
 
+begin_async_continuation
+````````````````````````
+
+::
+
+  sil-instruction ::= 'begin_async_continuation' '[throws]'? sil-type
+
+  %0 = begin_async_continuation $T
+  %0 = begin_async_continuation [throws] $U
+
+Begins a suspension of an ``@async`` function. This instruction can only be
+used inside an ``@async`` function. The result of the instruction is an
+``UnsafeContinuation<T>`` value, where ``T`` is the formal type argument to the
+instruction, or an ``UnsafeThrowingContinuation<T>`` if the instruction
```

Comment:
Random thought: should the argument of a lowered `Unsafe*Continuation` type be a formal type or a lowered type?  It's a lot easier to implement it if it's a lowered type, since otherwise we need a continuation-reabstraction operation (which can be done by allocating a context in the continuation, although then we have to deal with the possibility of code resuming a less-abstracted continuation — we need to free the other continuations).  And we can't take advantage of a better abstraction pattern on the application side unless we use custom lowering for the call to `resume`, since the builtin call would be hidden from SILGen behind code with an opaque pattern.  So making it a lowered type has real costs that we end up paying unconditionally.  But the decision here becomes ABI, so it's something we need to consider carefully.

Reply:
We could also just say that resume values are always maximally abstracted, which avoids the need to deal with reabstraction and the representational complexities of dealing with another special-case lowered generic parameter type. To me, the benefits of being able to use a better abstraction pattern across resume points seem pretty marginal.

Reply:
Since it's restricted to places that we need custom continuations, that's probably true, yeah.